packages/cursorless-vscode/src/keyboard/grammar/WorkQueue.ts

```diff
@@ -11,20 +12,15 @@ export class WorkQueue<T> {
   push(...items: T[]) {
     for (const item of items) {
       if (!this.items.has(item)) {
-        this.items.set(item, false);
+        this.items.add(item);
+        this.queue.push(item);
       }
     }
   }
 
   *[Symbol.iterator]() {
-    // while there are still items that haven't been yielded
-    while ([...this.items.values()].some((value) => !value)) {
-      // find any item that hasn't been yielded yet
-      const currentItem = [...this.items.entries()].find(
-        ([_, visited]) => !visited,
-      )![0];
-      this.items.set(currentItem, true);
-      yield currentItem;
+    while (this.queue.length > 0) {
+      yield this.queue.shift()!;
```

Comment:
do you want to remove item from items? (that is, do you want to be able to re-add an item for processing a second time after it has already been processed a first time?)

Reply:
To be fair, there are situations in which you want to re-use the work queue, and the structure of the problem guarantees that it is safe to remove from items. But fair enough. :P

Reply:
No I mean the only reason I created this class is because I was infinite looping due to recursion in the grammar causing me to visit the same node infinitely many times

Maybe I should rename this class to make it clear that's what it's for. `UniqueWorkQueue`? `IdempotentWorkQueue`?

Reply:
nah, it's fine. stet

Reply:
Unless you object, I do think I will rename it. The uniqueness thing is critical, and would be annoying to debug if someone accidentally removes that property

Reply:
go for it